src/lv_font/lv_font_loader.c

```diff
@@ -0,0 +1,680 @@
+/**
+ * @file lv_font_loader.c
+ *
+ */
+
+/*********************
+ *      INCLUDES
+ *********************/
+
+#include <stdint.h>
+#include <stdbool.h>
+
+#include "../lvgl.h"
+#include "../lv_misc/lv_fs.h"
+#include "lv_font_loader.h"
+
+#if LV_USE_FILESYSTEM
+
+
+/**********************
+ *      TYPEDEFS
+ **********************/
+typedef struct {
+    lv_fs_file_t * fp;
+    int8_t bit_pos;
+    uint8_t byte_value;
+} bit_iterator_t;
+
+typedef struct font_header_bin {
+    uint32_t version;
+    uint16_t tables_count;
+    uint16_t font_size;
+    uint16_t ascent;
+    int16_t descent;
+    uint16_t typo_ascent;
+    int16_t typo_descent;
+    uint16_t typo_line_gap;
+    int16_t min_y;
+    int16_t max_y;
+    uint16_t default_advance_width;
+    uint16_t kerning_scale;
+    uint8_t index_to_loc_format;
+    uint8_t glyph_id_format;
+    uint8_t advance_width_format;
+    uint8_t bits_per_pixel;
+    uint8_t xy_bits;
+    uint8_t wh_bits;
+    uint8_t advance_width_bits;
+    uint8_t compression_id;
+    uint8_t subpixels_mode;
+    uint8_t padding;
+} font_header_bin_t;
+
+typedef struct cmap_table_bin {
+    uint32_t data_offset;
+    uint32_t range_start;
+    uint16_t range_length;
+    uint16_t glyph_id_start;
+    uint16_t data_entries_count;
+    uint8_t format_type;
+    uint8_t padding;
+} cmap_table_bin_t;
+
+
+/**********************
+ *  STATIC PROTOTYPES
+ **********************/
+static bit_iterator_t init_bit_iterator(lv_fs_file_t * fp);
+static bool lvgl_load_font(lv_fs_file_t * fp, lv_font_t * font);
+int32_t load_kern(lv_fs_file_t * fp, lv_font_fmt_txt_dsc_t * font_dsc, uint8_t format, uint32_t start);
+
+static int read_bits_signed(bit_iterator_t * it, int n_bits, lv_fs_res_t * res);
+static int read_bits(bit_iterator_t * it, int n_bits, lv_fs_res_t * res);
+
+
+/**********************
+ *      MACROS
+ **********************/
+
+/**********************
+ *   GLOBAL FUNCTIONS
+ **********************/
+
+/**
+ * Loads a `lv_font_t` object from a binary font file
+ * @param font_name filename where the font file is located
+ * @return a pointer to the font or NULL in case of error
+ */
+lv_font_t * lv_font_load(const char * font_name)
+{
+    bool success = false;
+
+    lv_font_t * font = lv_mem_alloc(sizeof(lv_font_t));
+    memset(font, 0, sizeof(lv_font_t));
+
+    lv_fs_file_t file;
+    lv_fs_res_t res = lv_fs_open(&file, font_name, LV_FS_MODE_RD);
+
+    if(res == LV_FS_RES_OK) {
+        success = lvgl_load_font(&file, font);
+    }
+
+    if(!success) {
+        LV_LOG_WARN("Error loading font file: %s\n", font_name);
+        lv_font_free(font);
```

Comment:
Could we add a comment here explaining how this works for a future developer? My understanding from reading the code is that every successfully allocated structure is immediately zeroed - that way any pointers inside can be checked for being non-NULL and then freed.

Reply:
Yes. You are right, I comment will be good here.

Reply:
I added some comments here and and a more complete one before the static function `lvgl_load_font`, please see if it's ok.

Reply:
Looks good.